### ISSUE TITLE
Design API

### ISSUE THREAD

**Issue by lthibault** [lthibault/gortc#1]
As a completely arbitrary starting point, I suggest we try to stick as close to the JavaScript API as possible, while making minor changes to keep our code-base idiomatic.  Specifically, I thought we could replace callback functions with generator functions that read from an event channel.

As such, something like `RTCPeerConnection.onicecandidate` would be implemented as follows (note the idiomatic function name):

```Go
func (pc *RTCPeerConnection) OnICECandidate() {
    return <-pc.chICEcandidates
}
```
Thoughts?

**Comment by Mercor** [comment]
Using Channels as CallBacks sound ok.

The signalling Part should be exposed as interface so everyone could implement their own. But there should be also an implementation using websockets so you could easily interact with Websites.

**Comment by donovanhide** [comment]
I'd say just follow the standard as much as possible. It seems a lot of the data structures are very well documented:
https://www.w3.org/TR/webrtc/
Maybe a first good step is to build out all the structs and swap enums for const's with appropriate types and see if a natural API presents itself. Happy to do that first step :-)

**Comment by Mercor** [comment]
Always a good idea to start with the data structures, the rest follows naturally :-)

**Comment by lthibault** [comment]
> Maybe a first good step is to build out all the structs and swap enums for const's with appropriate types and see if a natural API presents itself. Happy to do that first step :-)

@donovanhide that sounds good to me!  Thanks for offering to do that part; please go ahead and I'll work on the websocket signaling in the meantime :)

> The signalling Part should be exposed as interface so everyone could implement their own. But there should be also an implementation using websockets so you could easily interact with Websites.

Agreed on both counts.  Having a reference implementation is important, I think, and websockets are probably what most people will want to use.

With that said, here's my proposal for a `SignalingChannel` interface.  I would like to propose a minor departure from the established API, however; since Go isn't callback-oriented like browser JS, I think it makes sense to nix the `onmessage` method in favor of `OnICECandidate` and `OnOffer`.

What are your thoughts?

```Go
type SignalingChannel interface {
    Send([]byte) error
    OnICECandidate() ICECandidate
    OnOffer() RTCPeerOffer
}
```

Have I overlooked anything?

**Comment by Mercor** [comment]
I think the SignallingChannel should use the ReadWriter (or at least the Writer) Interface, somthing like this:

```
type SignalingChannel interface {
    Writer
    OnICECandidate() ICECandidate
    OnOffer() RTCPeerOffer
}
```

so one could use the Channel everywhere a Writer is required.

**Comment by lthibault** [comment]
@Mercor agreed, thanks

**Comment by Mercor** [comment]
When i think over it, it semms to me that the signalingchannel just needs to have an reader and a writer, or ?
```GO
type SignalingChannel interface {
    Writer
    Reader
}
```
Because the one implementing the Interface doesnt wat to deal with ICE and SDPs i think, he is just providing a mechanism to send and receive data. Or am i wrong ?

Out webRTC Lib wants to read and to write to the channel, thats all. When there is an ICE Candidate, the callback should be called by another interface, something like 

```GO
type WebRTCHandler interface {
    OnICECandidate() ICECandidate
    OnSDP() SDP
    OnOffer() RTCPeerOffer
    CreateAnswer(SDP) SDP
}
```

What do you think ?

**Comment by santokum** [comment]
@Mercor - i agree - signalingchannel should be an abstract. ICE/offer-answer SDPs are all common denominator for whichever signaling mechanism is in play

**Comment by donovanhide** [comment]
There is no defined standard for signalling, so a application might use anything they want, such as websockets, eventsource, plain old HTTP GET (streaming with HTTP 2 :-)) or anything else, so perhaps this shouldn't be part of the package at all?

I found this set of slides to be a useful read and selection of links:

http://io13webrtc.appspot.com/

Just working through the spec now, getting familiar with all the data structures. Will push a fork over the next day or two with an initial stab at getting them into idiomatic Go.

**Comment by santokum** [comment]
Yes there is no mandate on signaling - if we are designing this from a purely library point of view then we don't need to worry about actual signaling mechanism itself, any native client can consume it irrespective of signaling channel used. However, if we are considering designing something similar to janus gateway then we need to look at transports/js api's..etc

**Comment by donovanhide** [comment]
@santokum A janus-like gateway in Go sounds like it should be another package (with a main function) hosted in another repo that imports this one, as well as gorilla/websockets or whatever else it needs to operate.
 
One thing not yet discussed is dependencies on SCTP and DTLS and whether the existing packages are sufficient.

**Comment by santokum** [comment]
Good point @donovanhide - for datachannel we need to support DTLS/SCTP  https://tools.ietf.org/html/draft-ietf-rtcweb-data-channel-13 and https://tools.ietf.org/html/draft-ietf-rtcweb-data-protocol-09

**Comment by lthibault** [comment]
@Mercor I'm not sure I quite grasp the distinction between `WebRTCHandler` and the original `SignalingChannel` proposal.  I get the sense that we're suggesting the same thing, and that the distinction is purely in the name.  If that's the case, then I have a slight preference for `SignalingChannel` as a name (I find it more descriptive, personally).

Perhaps the distinction you're making is that `SignalingChannel` / `WebRTCHandler` should have no opinion on the transmission interface, and only expose event-handlers?  That certainly seems reasonable.

With that said, I strongly feel that our pseudo-callbacks (i.e. `On*` methods) should return a `<-chan` rather than the struct itself, as this will allow us to use the `range` keyword when responding to these events.  To reiterate, I'm advocating `OnSDP() <-chan SDP` rather than `OnSDP() SDP`.

@donovanhide, @santokum (welcome!): I'm of the opinion that this package should at least have a simple working example in a `main` package.  A full-fledged Janus-like gateway as a separate project is, of course, a great idea, but I do think it's important to have an out-of-the-box working example for people.

>for datachannel we need to support DTLS/SCTP

I found this:  https://bitbucket.org/cyberroadie/go-sctp

It's BSD-licensed, so maybe we could just use that, or at least borrow heavily?  I'm of the opinion that "native Go" doesn't imply "reinvent the wheel"! ;)

Since we're sharing useful links, I'd also like to submit [this one](http://chimera.labs.oreilly.com/books/1230000000545/ch18.html)

>Just working through the spec now, getting familiar with all the data structures. Will push a fork over the next day or two with an initial stab at getting them into idiomatic Go.

Looking forward to it!

**Comment by donovanhide** [comment]
These days I tend to not export channels from packages, it can get complicated when error conditions need to be reported or timeouts handled as well as the actual asynchronous event that may or may not arrive.

https://talks.golang.org/2013/bestpractices.slide#25

**Comment by Mercor** [comment]
With SignalingChannel / WebRTC Händler just mean not to mix signalling and
webrtc related stuff.
Right, using channels in a public API is kind of smelly...
But i also think there should be a working signalling Solution that just
works out of the box.
Am 31.01.2016 13:51 schrieb "Donovan Hide" <notifications@github.com>:

**Comment by lthibault** [comment]
@donovanhide @Mercor  Might we consider a handler-based approach instead?

I suppose it's a minor issue, but forcing developers write a loop in order to call `On*` feels wrong to me.  How would you feel about:

```Go
type WebRTCHandler interface {
    HandleICECandidate(func(ICECandidate))
    // ...
```

This would allow for something like:

```Go
func (s SomeWebRTCHandlerImplementation) HandleICECandidate(h func(ICECandidate)) {
    for c := range s.chICE {
        h(c)
    }
}
```

**Comment by donovanhide** [comment]
I personally like the gorilla/websocket API of NextReader/Writer:

https://godoc.org/github.com/gorilla/websocket#Conn.NextReader

Not sure why you would want some client code to handle the ICECandidate event? I'd imagine client software would want to configure a peer, connect and check for errors and then just read and write data to it, much like you would a websocket client. The SDP handshaking and setting up the different levels of protocol (UDP/SCTP/DTLS) is the meat of what this library should handle internally, in my opinion. Am I missing something?

**Comment by lthibault** [comment]
>Am I missing something

No, no, you're not -- it's me, and I now feel quite silly.  My thinking got muddled from studying too many JS callbacks.  We're on the same page!

/hangs head in shame...

**Comment by donovanhide** [comment]
@lthibault Don't hand your head in shame :-) 

I agree that the spec itself has a very browser-centric and JS world view and probably a lot more of the nuts and bolts seems to be exported in the browser implementations. We probably don't need to export any way near as many of the data structures as has been done by the browsers, but only time will tell :-)

I think gorilla/websocket is a good battle-tested package that we should aim to emulate it as much as possible.

**Comment by Mercor** [comment]
Yes, the gorilla websocket implementation is a good starting point. Lets try their approach....
I also think that the internals like ICE and SDP should not be handled by the "upper layer", but there should be an option to do so...

**Comment by lthibault** [comment]
Yep, :+1: for gorilla as well, as well as for exposing internals via consistent interfaces.

**Comment by lthibault** [comment]
Given that it's Friday and I'm already looking forward to the weekend, I thought I'd check in real quick!

@donovanhide How is the initial push coming along?  Can I help with anything?

On an unrelated note, how do you guys feel about setting up a slack channel for this project?  It might be worthwhile to keep coordination details out of the public issue tracker.  I'd be happy to set that up.

**Comment by donovanhide** [comment]
@lthibault Hi, sorry for the lack of an update. Went down the rabbit hole and came to the conclusion that most of the building blocks are currently missing to build any usable WEBRTC implementation in pure Go. Bindings to openssl/boringssl for DTLS, usrsctp for SCTP over DTLS over UDP, and libnice for ICE would be one route.

However, I'm currently hacking together a barebones Go DTLS package from this starting point:

https://boringssl.googlesource.com/boringssl/+/63fa118f3a3d065adb6ca17227bb9e407ffadccb/ssl/test/runner

As detailed in the "Testing" paragraph here:

https://www.imperialviolet.org/2015/10/17/boringssl.html

Mostly for my own education on DTLS and UDP in general. In other words, I think this project is probably a bit premature in its conception :-)

**Comment by lthibault** [comment]
@donovanhide No worries -- I figured the workweek had gotten the best of you ;)

I sort of expected this to be the case:  why don't we just use cgo for the time being and then replace those pieces later (is that what you were getting at?) ?  Having a native Go API is a bit step in the right direction, IMHO.
